### DISCUSSION TITLE
Feature request: Incrementally updating FTS index

### DISCUSSION THREAD

**Discussion by vimota** [duckdb/duckdb#5352] [General]
Hi 👋  First of all, Duckdb and it's FTS extension is awesome! 

The current extension requires rebuilding the FTS reverse-index on any data-change, which for my dataset takes 5min+. For very infrequent changes that's not too bad, but for any sort of small somewhat-frequent changes the index becomes stale quite quickly. It would be awesome if the FTS index was able to update incrementally on inserts to an indexed table (or you could manually trigger an index update that was only on the new data). This would be super useful for dashboard and webapp-style applications.

**Comment by vimota** [comment]
It's possible that I could reverse-engineer how FTS stores the index and hackily append to it indices for the new data ?

**Comment by lnkuiper** [comment]
You can see how the fts index is built in our source code under `extension/fts`. It's just SQL queries. I am sure this indexing time can be improved, as I made the extension a long time ago.

**Reply by jez9999** [reply]
The problem isn't really the indexing time, it's the very fact that the whole index needs to be recreated from scratch.  Couldn't the implementation be modified to have the option of automatically adding to/updating/deleting from the index when the underlying column is added to/updated/deleted from?

**Reply by lnkuiper** [reply]
The FTS extension is purely SQL-based. You can have a look at it [here](https://github.com/duckdb/duckdb_fts/blob/main/extension/fts/fts_indexing.cpp).

BM25 uses corpus-wide statistics like term frequency, document frequency, and average document length. If the data is updated, these statistics must be updated as well. If DuckDB would have `TRIGGER`s, this could be automated purely using SQL, just like [SQLite's FTS extension](https://stackoverflow.com/questions/53733572/do-sqlite-fts-tables-need-to-be-manually-populated/53751620#53751620). We could bake it into the FTS extension so that these wouldn't have to be created manually.

We currently have no plans to implement `TRIGGER` (or something similar), although I could see us adding it in the future if it becomes a higher priority for us.

**Reply by jez9999** [reply]
Yeah.  Bit of a shame really because the inability to auto-update the index was a dealbreaker for me and I had to use the SQLite FTS instead (which IMHO is rather inferior because it requires the mess of moving columns to a whole separate virtual table rather than just applying indexes to what you want to index)

**Reply by hpvd** [reply]
imho this is a general topic, not only for FTS index but for all indexes used.

Lots of the speed possibilities results from indexes, so they should have to be used as often as possible...
To make this possible, they have to be up-to-date, which is not really possible in many use-cases without relying on automatic and incremental updates...

Maybe one day one can provide
- the needed mechanism incl triggers and
- some kind of architecture guide to implement this for any kind of index...

Just to give 2 examples what index can do:
1. with index active, classic relational dbs (MariaDB/Mysql) can even outperform graph dbs (neo4j) in typical graph usecases `with recursive` e.g. see task 5 in https://www.mdpi.com/2076-3417/12/13/6490

2. In the OLAP space, Pinot is the golden sample to benefit from Indexes in a stunning way:
see many benchmarks with and without indexes:
https://startree.ai/resources/what-makes-apache-pinot-fast-chapter-ii

=> maybe we should open an issue on enabling & standardize automatic, incremental Index updates?

**Reply by hpvd** [reply]
opened a super set of this (feature requests should start as discussions in this project)
**enabling & standardize automatic, incremental Index updates** https://github.com/duckdb/duckdb/discussions/15291
plz up vote it if you like to give it some visibility

**Comment by pengfeng** [comment]
By any chance this feature will be picked up any time soon? Or nobody has checked it since last update?

Any reason that it might be hard? Thanks!
